### ISSUE TITLE
SC2086: Double quote while using set

### ISSUE THREAD

**Issue by sahil-lakhwani** [koalaman/shellcheck#1284]
Rule Id: SC2086
shellcheck -V : 0.5.0

I am using [this compliance script](https://secscan.acron.pl/centos7/6/2/6).

shellcheck complains at the below line:
`set -- $p`

`error: SC2086: Double quote to prevent globbing and word splitting.`

The problem is that if I put a double quote around `$p`, the set command, I guess, interprets in a different way. Hence, the value of `$1` is computed same as of `$p` instead of the first instead of the first token from `$p`.

**Comment by matthewpersico** [comment]
Correct. With quotes, $p is one string, with spaces embedded. The solution is to add a shellcheck disable directive at the line above; that’s what they are for.

The REAL solution is to change the assignment to $p so that $p is an array. I am surprised that shellcheck didn’t complain about that assignment.

**Comment by koalaman** [comment]
In an `array=( $p )` you would have gotten a better warning "Quote to prevent word splitting, or split robustly with mapfile or read -a."

Maybe `set -- $p` should have a similar warning when the shell is bash, and be ignored in sh.

**Comment by sahil-lakhwani** [comment]
@matthewpersico I am not sure how would `set` work on an array then.

**Comment by matthewpersico** [comment]
```bash
declare -a p
readarray -t p < <(echo "$PATH" | sed -e 's/::/:/' -e 's/:$//' -e 's/:/\n/g')
set -- "${p[@]}"
```
Things to note:
1) The naive assignment ```p=($(echo $PATH | sed -e 's/::/:/' -e 's/:$//' -e 's/:/ /g'))``` fails any time an individual element of the PATH has a space in it, usually in Windows. Using the readarray properly keeps such paths as one element.
2) Furthermore, the change of : to \n instead of ' ' because readarray is line-based. It becomes more obvious when you lookup up readarray in the bash man page and see that it is just a synonym for 'mapfile'.

**Comment by sahil-lakhwani** [comment]
@matthewpersico Got it. Thanks for the insights.

**Comment by vapier** [comment]
should close this bug then ?  sounds like shellcheck is correct and the warning is WAI.

**Comment by vapier** [comment]
the linter is correct: the code written `set -- $p` undergoes globbing/splitting and needs quoting.  if you don't want it to warn, disable the warning in your code.

**Comment by philippgl** [comment]
I know, that `set -- $p` does the splitting and globbing. I think, this is the main reason to do the `set -- $p` in the first place. `set -- $p` is used to assign multiple variables, right?
So for the only (?) legitimate use of `set --`, I always have to disable the warning?
If you do `set --"$p"`, it just assigns the named variable `$p` to `$1`.

**Comment by philippgl** [comment]
Can I disable this w/o changing the file itself?
I do not want to clutter my source code with linter specific stuff. Probably I also need to do this for each linter?
The only option would be to pass the `--exclude=SC2056`, which would disable it for the whole file, right?

**Comment by vapier** [comment]
correct, `--exclude` affects the whole file, but does not require modification of it
